go/logger/buffer.go

```diff
@@ -0,0 +1,149 @@
+		sentinelCh: make(chan struct{}, 1),
+		timer:      time.NewTimer(0),
+		shutdownCh: make(chan struct{}),
+	}
+	go func() {
+		for {
+			select {
+			case <-t.timer.C:
+				<-t.sentinelCh
+				t.C <- struct{}{}
+			case <-t.shutdownCh:
+				t.timer.Stop()
+				return
+			}
+		}
+	}()
+	return t
+}
+
+func (t *triggerableTimer) ResetIfStopped(d time.Duration) {
+	if d == 0 {
+		return
+	}
+	select {
+	case t.sentinelCh <- struct{}{}:
+		t.timer.Reset(d)
+	default:
+	}
+}
+
+func (t *triggerableTimer) Close() {
+	close(t.shutdownCh)
+}
+
+type autoFlushingBufferedWriter struct {
+	lock           sync.RWMutex
+	bufferedWriter *bufio.Writer
+	backupWriter   *bufio.Writer
+
+	frequency    time.Duration
+	timer        *triggerableTimer
+	shutdown     chan struct{}
+	doneShutdown chan struct{}
+}
+
+var _ io.Writer = &autoFlushingBufferedWriter{}
+
+func (writer *autoFlushingBufferedWriter) backgroundFlush() {
+	for {
+		select {
+		case <-writer.timer.C:
+			// Swap out active and backup writers
+			writer.lock.Lock()
+			writer.bufferedWriter, writer.backupWriter = writer.
+				backupWriter, writer.bufferedWriter
+			writer.lock.Unlock()
+
+			writer.backupWriter.Flush()
+		case <-writer.shutdown:
+			writer.timer.shutdownCh <- struct{}{}
+			writer.bufferedWriter.Flush()
+			// If anyone is listening, notify them that we are done shutting down.
+			select {
+			case writer.doneShutdown <- struct{}{}:
+			default:
+			}
+			return
+		}
+	}
+}
+
+// NewAutoFlushingBufferedWriter returns an io.Writer that buffers its output
+// and flushes automatically after `flushFrequency`.
+func NewAutoFlushingBufferedWriter(baseWriter io.Writer,
+	flushFrequency time.Duration) (w io.Writer, shutdown chan struct{}, done chan struct{}) {
+	result := &autoFlushingBufferedWriter{
+		bufferedWriter: bufio.NewWriter(baseWriter),
+		backupWriter:   bufio.NewWriter(baseWriter),
+		frequency:      flushFrequency,
+		timer:          newTriggerableTimer(flushFrequency),
+		shutdown:       make(chan struct{}),
+		doneShutdown:   make(chan struct{}),
+	}
+	go result.backgroundFlush()
+	return result, result.shutdown, result.doneShutdown
+}
+
+func (writer *autoFlushingBufferedWriter) Write(p []byte) (int, error) {
+	// The locked resource here, the pointer bufferedWriter, is only being read
+	// even though this function is Write.
+	writer.lock.RLock()
+	defer writer.lock.RUnlock()
+
+	n, err := writer.bufferedWriter.Write(p)
+	if err != nil {
+		return n, err
+	}
+	writer.timer.ResetIfStopped(writer.frequency)
+
+	return n, nil
+}
+
+// EnableBufferedLogging turns on buffered logging - this is a performance
+// boost at the expense of not getting all the logs in a crash.
+func EnableBufferedLogging() {
+	writer, shutdown, done := NewAutoFlushingBufferedWriter(ErrorWriter(), loggingFrequency)
+	stdErrLoggingShutdown = shutdown
+	stdErrLoggingShutdownDone = done
+	logBackend := logging.NewLogBackend(writer, "", 0)
+	logging.SetBackend(logBackend)
```

Comment:
Are you sure this works as you expect it to? It seems to me that this would change all logging backends, not just the one for KBFS.

Reply:
not if this is only called from KBFS binaries, right?

Reply:
But everything is one binary on mobile.

Reply:
oh you're right!

Reply:
I forgot that everything is one binary on mobile. Should I go for the slightly bigger refactor to cut the global variables out entirely?

Reply:
that's why I mentioned that maybe it would be easier to special case logger on core side

because from what I observed, we always access logger through this variable:
https://github.com/keybase/client/blob/master/go/libkb/globals.go#L49

so we have our own global, in a way

Reply:
I don't want to cause any more work.  As @zapu says we can turn it off in the service if we want.  I just wanted the flexibility to do that, previous PR I believe had it always on...thanks!

Reply:
This PR has it on in the single process on mobile by default. Are you okay with this going in like that, or should we wait until you modify the service / should I modify the service to use a specific log backend?